src/instruction_graph_generator.cc

```diff
@@ -0,0 +1,2022 @@
+		assert(m_system.memories[mid_a].copy_peers[mid_a]);
+		for(memory_id mid_b = mid_a + 1; mid_b < m_system.memories.size(); ++mid_b) {
+			assert(m_system.memories[mid_a].copy_peers[mid_b] == m_system.memories[mid_b].copy_peers[mid_a]
+			       && "system_info::memories::copy_peers must be reflexive");
+		}
+	}
+#endif
+
+	batch init_epoch_batch;
+	m_idag->begin_epoch(task_manager::initial_epoch_task);
+	const auto init_epoch = create<epoch_instruction>(init_epoch_batch, task_manager::initial_epoch_task, epoch_action::none, instruction_garbage{});
+	if(m_recorder != nullptr) { m_recorder->record(epoch_instruction_record(*init_epoch, command_id(0 /* or so we assume */))); }
+	m_last_epoch = init_epoch;
+	flush_batch(std::move(init_epoch_batch));
+
+	// The root collective group (aka MPI_COMM_WORLD) already exists, but we must still equip it with a meaningful last_host_task.
+	m_collective_groups.emplace(root_collective_group_id, init_epoch);
+}
+
+void impl::notify_buffer_created(const buffer_id bid, const range<3>& range, const size_t elem_size, const size_t elem_align, allocation_id user_aid) //
+{
+	const auto [iter, inserted] =
+	    m_buffers.emplace(std::piecewise_construct, std::tuple(bid), std::tuple(range, elem_size, elem_align, m_system.memories.size()));
+	assert(inserted);
+
+	if(user_aid != null_allocation_id) {
+		// The buffer was host-initialized through a user-specified pointer, which we consider a fully coherent allocation in user_memory_id.
+		assert(user_aid.get_memory_id() == user_memory_id);
+		const box entire_buffer = subrange({}, range);
+
+		auto& buffer = iter->second;
+		auto& memory = buffer.memories[user_memory_id];
+		auto& allocation = memory.allocations.emplace_back(user_aid, nullptr /* alloc_instruction */, entire_buffer, buffer.range);
+
+		allocation.track_atomic_write(entire_buffer, m_last_epoch);
+		buffer.track_original_write(entire_buffer, m_last_epoch, user_memory_id);
+	}
+}
+
+void impl::notify_buffer_debug_name_changed(const buffer_id bid, const std::string& name) { m_buffers.at(bid).debug_name = name; }
+
+void impl::notify_buffer_destroyed(const buffer_id bid) {
+	const auto iter = m_buffers.find(bid);
+	assert(iter != m_buffers.end());
+	auto& buffer = iter->second;
+
+	batch free_batch;
+	for(memory_id mid = 0; mid < buffer.memories.size(); ++mid) {
+		auto& memory = buffer.memories[mid];
+		if(mid == user_memory_id) {
+			for(const auto& user_alloc : memory.allocations) {
+				// When the buffer is gone, we can also drop the user allocation from executor tracking.
+				m_unreferenced_user_allocations.push_back(user_alloc.aid);
+			}
+		} else {
+			for(auto& allocation : memory.allocations) {
+				const auto free_instr = create<free_instruction>(free_batch, allocation.aid);
+				if(m_recorder != nullptr) {
+					m_recorder->record(free_instruction_record(
+					    *free_instr, allocation.box.get_area() * buffer.elem_size, buffer_allocation_record{bid, buffer.debug_name, allocation.box}));
+				}
```

Comment:
The `if` blocks at every recording site are necessary to avoid evaluating the arguments to the constructor call of the record type if recording is disabled - otherwise we could have a `record<RecordType>(CtorParams &&...)` member function indeed. I don't see any other way to forcibly keep the recording code out of the non-recording execution flow at the moment.

Reply:
Thinking about this again, it could be possible to pass a lambda to `create`:
```c++
create<free_instruction>(free_batch, allocation.aid,
    [&] { return free_instruction_record(*free_instr, allocation.box.get_area() * buffer.elem_size, buffer_allocation_record{bid, buffer.debug_name, allocation.box}); });
```
... in this form, with some meta-programming shenanigans akin to `parallel_for` to extract the last argument.

Reply:
I've found a very DRY implementation for this that I'm pretty satisfied with now - see [5bdbc33](https://github.com/celerity/celerity-runtime/pull/249/commits/5bdbc331c0244017f4daa94d0991ddd6eac7b8ba).